expr/func_linearregression.go

```diff
@@ -0,0 +1,142 @@
+	s.endTargetAt = context.to
+
+	return context
+}
+
+func linearRegressionAnalysis(series models.Series, startSourceAt uint32, endSourceAt uint32) (float64, float64, bool) {
+	var n float64
+	var sumI float64
+	var sumII float64
+	var sumV float64
+	var sumIV float64
+
+	for i := sort.Search(len(series.Datapoints), func(i int) bool { return series.Datapoints[i].Ts >= startSourceAt }); i < len(series.Datapoints) && series.Datapoints[i].Ts <= endSourceAt; i++ {
+		// The index must be rebuilt from the timestamp because
+		// the points of the series do not include "missing" points.
+		index := float64((series.Datapoints[i].Ts - startSourceAt) / series.Interval)
+		value := series.Datapoints[i].Val
+
+		n++
+		sumI += index
+		sumII += index * index
+		sumV += value
+		sumIV += index * value
+	}
+
+	denominator := n*sumII - sumI*sumI
+	if denominator == 0 {
+		return 0, 0, false
+	}
+
+	factor := (n*sumIV - sumI*sumV) / denominator / float64(series.Interval)
+	offset := (sumII*sumV-sumIV*sumI)/denominator - factor*float64(startSourceAt)
+	return factor, offset, true
+}
+
+func (s *FuncLinearRegression) Exec(dataMap DataMap) ([]models.Series, error) {
+	loc, err := time.LoadLocation("") // todo, no idea what timezone to use here, utc is a reasonable default
+	if err != nil {
+		return nil, fmt.Errorf("failed to load location: %w", err)
+	}
+
+	now := time.Now()
+
+	defaultStartSourceAt := uint32(now.Add(-24 * time.Hour).Unix())
+	startSourceAt, err := dur.ParseDateTime(s.startSourceAt, loc, now, defaultStartSourceAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse 'startSourceAt' argument %q: %w", s.startSourceAt, err)
+	}
+
+	defaultEndSourceAt := uint32(now.Unix())
+	endSourceAt, err := dur.ParseDateTime(s.endSourceAt, loc, now, defaultEndSourceAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse 'endSourceAt' argument %q: %w", s.endSourceAt, err)
+	}
+
+	// todo update context.from and .to if needed?
+
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	results := []models.Series{}
+	for _, serie := range series {
+		factor, offset, forecast := linearRegressionAnalysis(serie, startSourceAt, endSourceAt)
+		if !forecast {
+			continue
+		}
+
+		datapoints := []schema.Point{}
+		{
+			var i uint32
+			for i = 0; i <= (s.endTargetAt-s.startTargetAt)/serie.Interval; i++ {
+				datapoints = append(datapoints, schema.Point{
+					Val: offset + (float64(s.startTargetAt)+float64(i)*float64(serie.Interval))*factor,
+					Ts:  s.startTargetAt + i*serie.Interval,
+				})
+			}
+		}
+
+		name := fmt.Sprintf("linearRegression(%s, %d, %d)", serie.Target, startSourceAt, endSourceAt)
+
+		newSeries := serie.Copy([]schema.Point{})
+		newSeries.Target = name
+		newSeries.Datapoints = datapoints
+		newSeries.Tags["linearRegressions"] = fmt.Sprintf("%d, %d", startSourceAt, endSourceAt)
+		newSeries.QueryPatt = name
+		newSeries.QueryFrom = s.startTargetAt
+		newSeries.QueryTo = s.endTargetAt
```

Comment:
I don't think this is right, but `timeShift` doesn't mess with these values. I'm not sure if timeShift is correct there...

Reply:
I think the Python implementation does set these values: https://github.com/graphite-project/graphite-web/blob/master/webapp/tests/test_functions.py#L4358